collect_app/src/test/java/org/odk/collect/android/external/FormUriActivityTest.kt

```diff
@@ -0,0 +1,158 @@
+import org.odk.collect.android.projects.CurrentProjectProvider
+import org.odk.collect.android.support.CollectHelpers
+import org.odk.collect.projects.InMemProjectsRepository
+import org.odk.collect.projects.Project
+import org.odk.collect.projects.ProjectsRepository
+import org.odk.collect.shared.strings.UUIDGenerator
+
+@RunWith(AndroidJUnit4::class)
+class FormUriActivityTest {
+    private lateinit var projectsRepository: ProjectsRepository
+    private lateinit var currentProjectProvider: CurrentProjectProvider
+    private val context = ApplicationProvider.getApplicationContext<Application>()
+    private val firstProject = Project.DEMO_PROJECT
+    private val secondProject = Project.Saved("123", "Second project", "S", "#cccccc")
+
+    @Before
+    fun setup() {
+        projectsRepository = InMemProjectsRepository()
+        currentProjectProvider = mock()
+
+        projectsRepository.save(firstProject)
+        projectsRepository.save(secondProject)
+        whenever(currentProjectProvider.getCurrentProject()).thenReturn(firstProject)
+
+        CollectHelpers.overrideAppDependencyModule(object : AppDependencyModule() {
+            override fun providesProjectsRepository(
+                uuidGenerator: UUIDGenerator?,
+                gson: Gson?,
+                settingsProvider: SettingsProvider?
+            ): ProjectsRepository {
+                return projectsRepository
+            }
+
+            override fun providesCurrentProjectProvider(
+                settingsProvider: SettingsProvider?,
+                projectsRepository: ProjectsRepository?
+            ): CurrentProjectProvider {
+                return currentProjectProvider
+            }
+        })
+    }
+
+    @Test
+    fun `When there are no projects then display alert dialog`() {
+        projectsRepository.deleteAll()
+
+        val scenario = ActivityScenario.launch(FormUriActivity::class.java)
+        onView(withText(R.string.no_projects_detected)).inRoot(isDialog()).check(matches(isDisplayed()))
+        onView(withId(android.R.id.button1)).perform(click())
+
+        assertThat(scenario.result.resultCode, `is`(Activity.RESULT_CANCELED))
+    }
+
+    @Test
+    fun `When there is project id specified in uri and it does not match current project id then display alert dialog`() {
+        val scenario = ActivityScenario.launch<FormUriActivity>(getIntent(secondProject.uuid))
+
+        onView(withText(R.string.wrong_project_selected_for_form)).inRoot(isDialog()).check(matches(isDisplayed()))
+        onView(withId(android.R.id.button1)).perform(click())
+
+        assertThat(scenario.result.resultCode, `is`(Activity.RESULT_CANCELED))
+    }
+
+    @Test
+    fun `When there is project id specified in uri and it matches current project id then start form filling`() {
+        Intents.init()
+
+        ActivityScenario.launch<FormUriActivity>(getIntent(firstProject.uuid))
+
+        Intents.intended(hasComponent(FormEntryActivity::class.java.name))
```

Comment:
We should probably check the intent details whenever we launch `FormEntryActivity` (like you do in `When everything is fine form filling is started with given data`).

Reply:
Every time check data and extras?

Reply:
Yeah I think so! Just to make sure we're definitely sending the right stuff through in each case.

Reply:
Ok, done.